src/main/kotlin/net/insprill/robotinsprill/form/FormHandle.kt

```diff
@@ -0,0 +1,200 @@
+
+import dev.kord.common.entity.ButtonStyle
+import dev.kord.common.entity.DiscordPartialEmoji
+import dev.kord.common.entity.InteractionType
+import dev.kord.common.entity.Permission
+import dev.kord.core.behavior.channel.createMessage
+import dev.kord.core.behavior.edit
+import dev.kord.core.behavior.interaction.respondEphemeral
+import dev.kord.core.behavior.interaction.updateEphemeralMessage
+import dev.kord.core.entity.Embed
+import dev.kord.core.entity.User
+import dev.kord.core.entity.component.ButtonComponent
+import dev.kord.core.entity.interaction.ComponentInteraction
+import dev.kord.core.entity.interaction.ModalSubmitInteraction
+import dev.kord.core.event.interaction.ComponentInteractionCreateEvent
+import dev.kord.core.event.interaction.ModalSubmitInteractionCreateEvent
+import dev.kord.core.event.message.MessageCreateEvent
+import dev.kord.core.on
+import dev.kord.rest.builder.component.ActionRowBuilder
+import dev.kord.rest.builder.message.EmbedBuilder
+import dev.kord.rest.builder.message.create.MessageCreateBuilder
+import dev.kord.rest.builder.message.create.embed
+import dev.kord.rest.request.KtorRequestException
+import kotlinx.datetime.Clock
+import net.insprill.robotinsprill.RobotInsprill
+import net.insprill.robotinsprill.configuration.BotConfig
+import java.net.MalformedURLException
+import java.net.URL
+import java.util.Collections
+
+class FormHandle(val robot: RobotInsprill) {
+
+    fun setupEventHandlers() {
+        robot.kord.on<ModalSubmitInteractionCreateEvent> {
+            handleForm(interaction)
+        }
+        robot.kord.on<ComponentInteractionCreateEvent> {
+            handleComponent(interaction)
+        }
+        robot.kord.on<MessageCreateEvent> {
+            if (message.author?.isBot == true) return@on
+            if (hasManageMessage(message.author ?: return@on)) return@on
+            if (robot.config.forms.any { it.channel == message.channelId } )
+                message.delete("Form only channel")
+        }
+    }
+
+    private suspend fun handleForm(interaction: ModalSubmitInteraction) {
+        if (interaction.type != InteractionType.ModalSubmit) return
+
+        val form = robot.config.forms.firstOrNull { it.name == interaction.modalId } ?: return // Kotlin :o
+        val invalids = form.fields.filter { it.isNumber == true }.filter {
+            try {
+                return@filter !it.range().contains( // just in case smashed in a few more currencies
+                    Integer.parseInt(interaction.textInputs[it.name]?.value?.replace(Regex("[.,$€£]"), ""))
+                )
+            } catch (e: NumberFormatException) {
+                return@filter true
+            }
+        }
+        val actionRow = makeInteractions(form)
+        val embed = makeEmbed(interaction, form, invalids)
+
+        val message: MessageCreateBuilder.() -> Unit = {
+            embed(embed)
+
+            if (invalids.isEmpty()) components.add(actionRow)
+            else content = "**Uh oh**"
```

Comment:
Is this possible to trigger? If so there should be a guard clause instead of this. If not, it can be removed.

Reply:
I am not quite sure what you mean here by "trigger." Could you elaborate on this?

Reply:
Sorry, I was referring to that if-statement.

Reply:
Not sure if 9bde37db matches what you mean. Could you check it out?